appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java

```diff
@@ -0,0 +1,143 @@
+package com.google.appinventor.client.editor.simple.components;
+
+import com.google.appinventor.client.ErrorReporter;
+import com.google.appinventor.client.Ode;
+import com.google.appinventor.client.editor.simple.SimpleEditor;
+import com.google.gwt.user.client.rpc.AsyncCallback;
+import com.google.gwt.user.client.ui.Image;
+
+import java.util.*;
+
+import static com.google.appinventor.client.Ode.MESSAGES;
+
+public class MockCSVFile extends MockNonVisibleComponent {
+  public static final String TYPE = "CSVFile";
+
+  private static final String PROPERTY_NAME_SOURCE_FILE = "SourceFile";
+
+  private String sourceFile;
+
+  private List<String> columnNames; // First row of the CSV contents
+  private List<List<String>> rows; // Parsed rows of the CSV file
+  private Set<CSVFileChangeListener> csvFileChangeListeners;
+
+  /**
+   * Creates a new instance of a CSVFile component whose icon is
+   * loaded dynamically (not part of the icon image bundle)
+   */
+  public MockCSVFile(SimpleEditor editor, String type, Image iconImage) {
+    super(editor, type, iconImage);
+
+    csvFileChangeListeners = new HashSet<CSVFileChangeListener>();
+  }
+
+  /**
+   * Changes the Source File property of the MockCSVFile.
+   * Upon change, the new file is imported and parsed, the
+   * contents are stored, and CSVFileChangeListeners are
+   * updated.
+   *
+   * @param fileSource  name of the new Source File
+   */
+  private void setSourceFileProperty(String fileSource) {
+    // Update the source file property & reset the
+    // columnNames property
+    this.sourceFile = fileSource;
+    columnNames = new ArrayList<String>();
+
+    // Check that the SourceFile property is a valid file
+    if (fileSource == null || fileSource.equals("")) {
+      // Invalid source file property: update CSVFileChangeListeners
+      // to notify of the change (columnNames now empty)
+      updateCSVFileChangeListeners();
```

Comment:
Consideration: updateCSVFileChangeListeners might be invoked at the beginning as well (e.g. in case reading fails)